app/src/main/java/rick_and_morty/ui/characters/composables/CharacterList.kt

```diff
@@ -0,0 +1,76 @@
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.lazy.LazyColumn
+import androidx.compose.foundation.lazy.items
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.collectAsState
+import androidx.compose.runtime.livedata.observeAsState
+import androidx.compose.ui.unit.dp
+import androidx.lifecycle.viewmodel.compose.viewModel
+import rick_and_morty.ui.characters.CharactersViewModel
+
+@Composable
+fun CharacterList(charactersViewModel: CharactersViewModel = viewModel(modelClass = CharactersViewModel::class.java)) {
+
+    val characters = charactersViewModel.characters.collectAsState(emptyList())
+    LazyColumn (
+        verticalArrangement = Arrangement.spacedBy(5.dp)
+    ) {
+
+        items(characters.value) {characters ->
+            CharacterRow(characterResultsDto = characters)
+        }
+        item {
+            LaunchedEffect(true) {
+                    charactersViewModel.getCharacters()
```

Comment:
So only the first page is loaded in the list?

Reply:
the logic of loading pages moved to the ViewModel, in the beginning it will load first page, as you scroll down second page loads

Reply:
Got it, but charactersViewModel.getCharacters() is called multiple times when scrolling up and down again with the same page number, resulting in duplicate characters in the list. You can simulate this by adding a delay after line 39 in `CharactersViewModel`, for example, `delay(10_000)`, you will see duplicate entries in the list.

Reply:
@pkliang I have tested it today, true it is called multiple times, but the list is not duplicated, it still remains the same as was

Reply:
Can you explain why the method with the same page id is called multiple times while the list has no duplicate entires? I think you better write a unit test for this case

Reply:
https://developer.android.com/jetpack/compose/side-effects#launchedeffect

@pkliang I understood is as even if LaunEffect called multiple times, it will change when it is recomposed with different keys.

Reply:
I remembered I can reproduce the bug last time, but I didn't have time to check the last change you made. Anyway,  calling the network API multiple times with the same request sounds not reasonable. But this is your pet project, it is up to you to optimise it.

Reply:
@pkliang I should definitely prevent it, my plan to do this with next PRs when I will save scrolling position.